### DISCUSSION TITLE
Improving the Tab Behavior

### DISCUSSION THREAD

**Discussion by ardeay** [zesty-io/manager-ui#1054] [Ideas]
Tabs are a feature of the Zesty manager that were put in place to create fast switching between two working environments, like a code view and a content editing view.  A tab is created for each experience (unique url) a user visits in Zesty, they stack left to right. When a tab is clicked, a user quickly jumps back to their previous working environments. Tabs were created because the developers witnessed that users would often keep multiple `web browser` tabs open to edit their zesty manager.  Keeping multiple tabs open is problematic because state changes to the data won't always be updated, which can create issues like version over written or updating stale data. Tabs were designed to behave like `web browser` tabs.

**Tabs get unruly quickly as their a so many views in any given Zesty manager.**

## We Want to Hear Your Feedback!

We have heard many complaints about tabs: too many, lose track, don't work with multi-lang, etc. 

**Please share your ideas on how to improve the Tab experience.** 

## Example of Tabs
Tabs are on the top of the interface
![image](https://user-images.githubusercontent.com/729972/168093743-2eadb08b-d892-4e1c-af23-3e311b53508b.png)
Tab bar view
![image](https://user-images.githubusercontent.com/729972/168095823-4d113a17-7090-48f4-9bb7-d07ccfd62824.png)
Close up
![image](https://user-images.githubusercontent.com/729972/168095861-2b266894-739f-404b-a2e9-930154a671fb.png)
Close up with an active tab
![image](https://user-images.githubusercontent.com/729972/168095901-3a58968c-5e70-4c78-9174-266fc85e601f.png)

**Comment by giseleblair** [comment]
A pin icon or some sort of action that would create a tab. Rather than having tabs stay put automatically. An auto-generated tab creates unnecessary clutter in the tab navigation that the nav itself becomes cumbersome to use.

**Reply by ardeay** [reply]
That's a cool idea! On how its implemented, would it be in the editing experience view, some sort of button top right or left that lets you pin? Or would it be a pin on the tab (but the tab would go away unless its pinned?)

**Reply by giseleblair** [reply]
I think a pin on the tab itself. Having the currently active tab is helpful to see what item you are currently on. So keeping the tab experience in the navigation itself makes sense to me

**Reply by ardeay** [reply]
What do you think of a solution where are always 3-5 `recent active tabs` plus any tabs that have been `pinned` in addition?

**Reply by giseleblair** [reply]
I think just the active tab and any pinned items would be good. If anything having 3-5 recent tabs would cause the same issues. If a user already has about 5 tabs pinned, then the recent tabs would cause clutter again

**Reply by ardeay** [reply]
Good points. I'm thinking through the implementation. What do you think of a max of 5 total tabs, with a max of 4 pinned tabs, given the 5th one is always active? At any given time there would be max 3 recent, if there was room.

How I'm imagining a solution like this is there would be no `x` (close) on the tab, just a pin icon, unpinning a pinned item would pop it off the queue unless it was the `active` tab. If a tab is not pinned, it falls of the queue as the user browses around.

**Comment by Katiemoser** [comment]
I honestly never use them, with them being dark with blue writing, I don't really notice them or think to use them in my workflow. I'm more inclined to navigate through the sidebar, so I agree with Gisele that it might be good to use it as a pin feature that people can save their most-used pages to easily navigate to.

**Comment by bltarkany** [comment]
A dropdown with a search, similar to VS code's `go to file` - where it shows the items pinned from Gisele's comment above and/or shows recently viewed files. And the ability to remove them from the dropdown via the `X`. Could possibly extend this out to include a search. The dropdown would keep the navigation free of clutter 
 
![image](https://user-images.githubusercontent.com/50595614/168134342-fcf4f459-5227-4417-b78c-4fa02610180f.png)

**Reply by ardeay** [reply]
That would be great for search.

**Comment by dhakimzadeh** [comment]
The problem with tabs

- NO SIDE SCROLLING, ever, side scrolling is the worst when not using a mac trackpad <--

-- when enough tabs are open to necessitate side scrolling, it becomes easier to just use the left columns again
--- one solution is to just limit tabs , but i dont like this

--- TAB GROUPS --- 
another option is to condense them into dropdowns under more global tabs like, CONTENT, MEDIA, SCHEMA, CODE.  

this way, if i open 3 code templates, 3 content forms, and jump to a data model, then jump to code, then after want to enter content on my form, i can immediately find the content form i was just on under my content tab dropdown, like tab groups with chrome, just more automatic,

groups would save me from clicking a multi item model then finding my item, i could just keep that in tab history under the content tab etc etc

the only problem here is: 
if I go to content, 
click on Articles, 
then click on the article i want to edit, 
then click on code and click on the code file i want to edit, 
when i go up to the dropdown for content, i would see Articles AND the specific article i want to edit

this is precisely what is causing so many tabs now, in reality, i dont need Articles to be there cluttering up the tabs, i just need my specific article -- only solve i can think of here is to manually choose what becomes a tab, but that doesnt feel like the best solve

what might work is limiting tabs, but also allowing pinning tabs like described above, but youd pin them to the tab group...

this way regardless of window width, your tabs are all there without sidescrolling to get to them ( global tabs could  just use the icons if the window gets too small)

**Comment by ardeay** [comment]
@dhakimzadeh for tab groups, are you imagining it like a traditional toolbar? 
![image](https://user-images.githubusercontent.com/729972/168313675-51c26884-1637-4148-ac40-a01dd09295e2.png)

**Reply by dhakimzadeh** [reply]
more like the mac dock with the open windows

![image](https://user-images.githubusercontent.com/59213/168324021-86d4ef2c-a6b1-424a-bfbd-d1aa7a3980f5.png)


although I do think a traditional toolbar might work for the far left nav

**Reply by ardeay** [reply]
ahh gotcha, so when a user would hover or an icon (say) schema on the left, it would give a flyout like this with latest opens?

**Reply by ardeay** [reply]
@agalin920 check this out

**Reply by dhakimzadeh** [reply]
yeah exactly, whether its top bar or left bar doesnt matter much. the mac dock is probably the best example, just swap in tabs for open windows :)

**Reply by ardeay** [reply]
In conclusion, the idea is when the user hovers over an app on the left, it has a fly out of most recent items visited there. This is good because if something in schema was worked on 6 clicks ago, it wont be in the tabs, but it will be seen in the most recent flyout. Making it so you are always 1 click away from your most recent schema edits, but without the clutter.

**Comment by jjnotjayjay** [comment]
@ardeay 

I love the Zesty platform as a whole!  But won't mince words below.

Tabs in their current form on Zesty are, essentially, unusable.  

**On a mature Zesty instance (with hundreds if not thousands of content items), as a user navigates between content items, the number of tabs continually increases, to the point where the number of tabs inevitably extends beyond the viewport's available width, leading to a scrollbar.**  (Because let's face it, it'd waste more time than it'd save to have to close a Zesty tab every time we switched views.). 

So for one example.  Say I worked on something 2 weeks ago, and its content view is thus now buried as a tab off the screen to the right.  I log into the Zesty instance to make some changes to unrelated content.  I open up a few different content items (opening up new tabs in the left-most portion of the tab group) and see them reflected.  I then have to make a change to the content item I edited 2 weeks ago.  The moment I open up the content item, the tab for that content item is opened up, auto-scrolling the tab nav to that item, and just like that, the tabs for all the other items are no longer in view.  If I wanted to get back to them, I'd have to scroll either in the tab bar (difficult as I develop on a keyboard and not a trackpad) or just scroll to the item in the content editor.  I imagine most if not all users simply scroll to the item in the content editor.

Naturally, it'd be a bit ridiculous to expect the user to keep track of the visibility and location of every tab that is currently up in the tab interface as the number of tabs continually increases, so for all intents and purposes I simply ignore the functionality.

**The best idea I believe I've had to improve the functionality as a whole is also one of the simplest: always open a new content item in the left-most space in the tab bar, unless the user opened the content item by clicking on its tab.**

Put another way, organize the tabs _always_ by most recently opened (by a means other than navigating via its tab).  So if I go to edit a new schema, a tab for the schema opens up in the left-most space.  I then open up an associated content item.  It's now in the left-most space, pushing the schema to the second space.  I make some changes to the content item and save.  I realize I need a new schema field.  I click the tab for the schema.  Both tabs remain where they are. I make the schema changes.  I navigate back to the content item by its tab.  I add the content.  I open up a new unrelated content item.  It's now in the left-most space, pushing the content and schema further to the right.

**I think this approach would solve most of the issues with tabs.  It eliminates the need for scrolling, and makes the behavior very predictable.  Users don't need to keep track of where anything is.** 

If they were organized like this, I think they'd be very useful and I'd actually use them.

**If this was combined with the ability to drag and drop a tab into a desired space, I think tabs would be phenomenal.**

**Reply by ardeay** [reply]
Thank you for the feedback @jjnotjayjay 

The left most loading for most recent makes sense. If you see the other user posts about pinned items, maybe pinned to right, most recent to left, the 3rd most recent pops off the stack. 

After reading every comment here, its starting to click that the solution will be a mixture of this feedback, which is great!

**Reply by jjnotjayjay** [reply]
I don't imagine I'd use pinned tabs, but another option would be to have all the pins in a dropdown that appears when the user hovers over a "pin" icon (on either side).

**Reply by shrunyan** [reply]
I agree with @jjnotjayjay suggestion. To me though it begs the question what is the more common use case;

- Recency. Quickly getting back to the last few views you had open. 
- Recall. Being able to bubble up(pin) a view so as you work it's ever present and can be quickly returned to.

**Comment by ardeay** [comment]
It's clear the current tab setup is poor and needs an improvement, based on your feedback I mocked up these concepts as a proposal for development. Comment or upvote if we should work these into production.

**Tab Pinning**
--
![image](https://user-images.githubusercontent.com/729972/168704355-680a49ce-2ee9-4e4b-b160-90a073028232.png)
![image](https://user-images.githubusercontent.com/729972/168704372-75a91b5f-1413-48ae-ad72-9edc52b15bf5.png)

ty @giseleblair for pin idea
ty @jjnotjayjay for left to right stack push

**App Recent Flyout**
--
![image](https://user-images.githubusercontent.com/729972/168705128-36fdf432-5280-4522-998b-4abfb43c4b08.png)

ty @dhakimzadeh for the flyout idea

**Reply by jjnotjayjay** [reply]
@ardeay Feedback on the proposals below:

**Tabs:** 
I'm not a huge fan of the number of tabs being that low (or even a fixed amount at all).  I think the number of tabs should continue to fill up the horizontal amount of the viewport - it just shouldn't go beyond it.  So perhaps the most 10 recent tabs are saved, and the number shown is viewport-dependent.

With similar reasoning, if a user hasn't pinned any tabs, there shouldn't be space dedicated to pinned tabs.  

Personally, I'd be completely fine without any pinned tab functionality since I imagine there wouldn't be a huge UX improvement over simply making it recently visited and on the left (after all, the user would be able to easily re-access that content item after their first visit), and fear it'd introduce unnecessary complexity.  I'm not against it though and could see it being useful, potentially (though not for any of the instances I work in) - I just worry that it might clutter or complicate the tab interface until the point where it's not particularly usable again.

**Flyout:**
I am personally strongly against this idea as it feels duplicative to me, if tabs are organized by most recently viewed.  I personally believe it would over-complicate the UI of the Zesty admin app, with very little marginal improvement.  At a minimum, I'd request this functionality be able to be turned off.

**Reply by shrunyan** [reply]
I concur on fly outs being undesired on the global app menu. Beyond @jjnotjayjay points, this could also just be interactively problematic. As you may intend to click on the global app menu link to load an app and then quickly mouse over to where you know that apps nav will be. But you will have triggered the fly out which would lay on top of that area in the view port possibly preventing you from clicking on the app nav link you need. Requiring you to hover away and then mouse back to the link. I think this would be extremely frustrating.

**Reply by ardeay** [reply]
True @shrunyan, the operating systems skirt that issue with input, on windows a right click gives you a flyout, on make a hold click. Both would be slick.

**Reply by ardeay** [reply]
@jjnotjayjay it would be easy to limit tabs to your view port with some simple division, making it not fixed to 3 but (width-available / tab-width)

**Comment by shrunyan** [comment]
**Discovery**, **Organization** or **Recency**

How are people perceiving the purpose of global tabs?

Is it about being able to **discover** items? Probably not, this would make more sense for the global search. Meaning you are wanting to work with a particular item and you need to find it amongst all the other items. This would also include a scenario where you are wanting to work on an item you have never viewed before. Therefore will not be recent and could be deep on a long list of items. 

Is it about **organizing** your workspace? Meaning you know you have to work on a web page which consists of a schema, content and code. You need to quickly jump back and forth between these views. Therefore you want these currently relevant views collected together.

Is it about **recency**? Being able to quickly hop back to the last few views you were working on. I could see this being the case in 2 scenarios. One when you are returning at a later point in time to the product, .e.g. The next working day, and you want to quickly return to the last place you were to begin working again. Another which is as you are working you want to be able to quickly return to a prior view. 

This was the original intent of the tabs. With both scenarios feeling like they are poorly served with the current experience.

My conclusion is that discovery is better served by a global search. Recency was the original intent but seems like it does not fit day-to-day workflows. **Organization sounds like the true underlying need.**

To this goal of organization. I suggest we combine a few of the already mentioned ideas into a cohesive solution.
1. All "tabs" are manually added. This creates a tab "group".
2. Each tab group allows for an "add" action. Which would add the current view to the group.
3. Allowing for users to create groups of views mapped to their organizational needs.

Overall the questions I am hearing we need to answer with any solution are the following;
- How do I get back to the prior view(s)?
- How do I find that item I was working on?
- How do I quickly switch between related views? E.g. content, schema, code
- How do I reorient myself when returning the next day?

E.g. 
![Master Spec Outline v 1 05-2021](https://user-images.githubusercontent.com/208819/168710825-75318a0e-4437-4ec0-b3ab-835f10f363aa.svg)

**Reply by shrunyan** [reply]
p.s. layout is for illustrative purposes. We would refine placement and edge cases if the concept is accepted.

**Reply by shrunyan** [reply]
I believe this solution will allow for each individual to create and delete arbitrary groups of relative work. 

For example 1 week you might make a "product announcement" group. Add a bunch items to the group and work on them through out the week. After the announcement is made you would delete the group. rinse and repeat.

My guess would be on average you would have 3 - 5 active groups. <- **Could definitely use some feed back on this assumption**. If the amount of active groups was far more it could put the experience back into the need for a horizontal scroll, making it undesirable.

**Reply by shrunyan** [reply]
One thing this proposed solution does not account for is "one offs". A single view you want to quickly return to that is not part of a logical group of work or that just on a regular basis needs to be visited. This could be accounted for if needed. Thoughts?

**Reply by jjnotjayjay** [reply]
Single views of 1-2 pages are probably 40-50% of the logins for some instances.

The only thing I wanted to definitely comment on is I'd like tabs to be created automatically.  It feels like it'd be a hassle to have to manually create a tab every time you want the ability to get back to something - especially if you aren't sure beforehand if you'll need to access it a second time.

**Reply by shrunyan** [reply]
@jjnotjayjay 
- Are the _single views of 1-2 pages_ the last few pages viewed on that instance? 
- Are they deep links within the instance and it's more about being able to quickly discover where those are? 
- What's the mindset when trying to find these pages?
  - You are working on a webpage(example.com/page) and need to edit it.
  - You been given edits and go straight to an instance and are not thinking in terms of a "webpage". What terms would you be thinking in? 
    - You need to edit this weeks blog post and you know where that is on the instance in question.
    - You need to find an item last published months ago and updated it?

**Reply by jjnotjayjay** [reply]
Marketer XYZ can't figure out how to change copy on ABC page (or is just super lazy).  I have to find the associated content item, make the change, send it to them for review/publish.  They _might_ come back with another change request for the same thing, might not.

Above is a common use case.  Generally it's not the last few pages viewed, generally it's older.  But I wouldn't think to manually add a tab for a change like that - but sometimes I do need to get back to it (i.e. they might say something like "hey the copy would look better if it was two lines not three, so can you pretty please remove the word Foo").

Not deeplink related.

Mindset is get the change done as quickly as possible.  Yes, I am thinking in terms of a webpage (almost always).  Could be either of those final two scenarios, where perhaps the latter is more likely since generally the content editor knows where something is if they need to change it frequently (i.e. weekly blog posts or something).

**Reply by shrunyan** [reply]
Thank you @jjnotjayjay this is very insightful. @ardeay this sounds like a use case best served by the Explorer feature. We should get another discussion going with regards to that feature.

The deeper need I am hearing is around **discovery**. Ultimately if marketer XYZ can not figure out how to change the copy that is the core problem we should be trying to address. I think the Explorer may answer that need. The outcome would hopefully be two fold.

1. marketers are more empowered and can easily self serve changes.
2. developer time is freed up by not needing to interrupt their flow to implement content changes.

Please let me know if I am off base @jjnotjayjay

**Reply by shrunyan** [reply]
> Yes, I am thinking in terms of a webpage (almost always) 

@jjnotjayjay if almost always thinking in terms of a webpage(i.e. URL). Would it be better to have a more robust global search in regards to searching by URL?

A work flow I am imagining is as follow;

1. Global search for `example.com/page-with-edit/`
2. Click into content item
3. Make content change
4. Click save & publish
5. Click accelerator link `Edit View` on content item
6. Make related code change
7. Click save & publish

**Reply by ardeay** [reply]
My opinion is the more the computer can do for you the better, which is why I dislike only groups since they require user input outside of a click. I do however like the ideas of a group option, where you could group a model/item/view or whatever you wanted, and name the group. Like a working tab or something of the likes. Some zesty instances have 60K content items, some have 50, very different use cases to consider. Bigger sites with power users, I imagine, would lean towards more control over the interface. Though the idea feels novel, and would be cool at first but would too much work to build up and tear down, which would leave users with dead tab groups. That is why I go back to automation building recent accessed along with user interaction to pin.

**Reply by jjnotjayjay** [reply]
Well, yes 40-50% for some instances is the use cases I outlined.  The other 50-60% of the time though, I'm developing new pages and would use tabs heavily if they were organized sensibly (i.e. most recent left), rather than opening things and have the tab bar horizontally scroll/jump around.

Discovery feels like it should be solved by the Global Search function.  The ability to find a string of copy via Global Search would be very useful, for example.

Sure, in a perfect world marketers find what they're looking for all the time, and choose to invest the effort to do things themselves.  Often time that simply doesn't happen - other times they can't be blamed for not finding something.  For example, the current Acorns instance for dot com has 214 different schemas, so I could see it being tough to find something once in a while.  (And that's leaving the "In-App" instance alone entirely, which is borderline unusable at times due to its size (even for me, it takes a lot of poking around).

In instances where my team or I have created everything personally, this is less of an issue (if an issue at all).  But there are instances that weren't carefully thought through during creation, which is where discovery becomes an issue.

Again, though, that's a completely separate conversation and I personally don't view tabs as a way to tackle the problem of discovery (at all).

**Reply by shrunyan** [reply]
We could possibly combine all of the feature concepts; _recent_, _pinned_, _groups_

Space is an important consideration as it is very constrained but fly outs may resolve that. Overload is another consideration. If it becomes a bar of icons(actions) that maybe overwhelming. 

For example
![spec_ manager-ui_tabs](https://user-images.githubusercontent.com/208819/168718444-6c38bd11-7724-4070-b7f9-2e9f67543e0d.svg)

**Reply by ardeay** [reply]
@shrunyan How are items added to a group? 

The combined one would likely make power users happy :)

**Reply by shrunyan** [reply]
I imagine on the group tab itself there would be an action which would add the current view(i.e. URL) to the group
